index.bs

```diff
@@ -1294,44 +1296,112 @@ A {{ReceiveStream}} has the following internal slots.
-stream (for QUIC, that message is an RESET_STREAM frame), the user agent MUST
-queue a task to run the following:
-      1. Let |stream| be the {{ReceiveStream}} object for which the reset
-         message was received.
-      1. Let |transport| be the {{WebTransport}}, which the |stream| was created
-         from.
-      1. Let |reason| be the value from the message from the remote side.
-      1. [=ReadableStream/Error=] |stream| with |reason|.
-
-The {{ReceiveStream}}'s <dfn>cancelAlgorithm</dfn> is:
-     1. Let |stream| be the {{ReceiveStream}} object.
-     1. Let |transport| be the {{WebTransport}}, which the |stream| was created
-        from.
-     1. Let |reason| be the first argument, if it is numeric, or 0.
-     1. If |stream| has received a RESET_STREAM frame, abort these steps.
-     1. [=In parallel=], [=send STOP_SENDING=] over |stream|'s
-        [=WebTransport stream=] with |reason| as the Application Protocol Error
-         Code.
+<div algorithm>
+
+To <dfn for="ReceiveStream">pull bytes</dfn> from a {{ReceiveStream}} |stream|, run these steps.
+
+1. Let |controller| be |stream|'s [=ReadableStream/[[controller]]=].
+1. Assert: |controller| is a {{ReadableByteStreamController}}.
+1. Let |transport| be |stream|'s [=ReceiveStream/[[Transport]]=].
+1. Let |internalStream| be |stream|'s [=ReceiveStream/[[InternalStream]]=].
+1. Let |promise| be a new promise.
+1. Let |view| be [=ReadableStream/current BYOB request view=] for |stream|.
+1. If |view| is null, set |view| to a [=new=] {{Uint8Array}} with an [=implementation-defined=]
+   size. If allocating the {{Uint8Array}} fails, return a promise [=rejected with=] a
+   {{RangeError}}.
+1. Let |bytes| be the byte sequence |view| represents.
+1. Assert: |bytes|'s length equals to |view|.\[[ByteLength]].
+1. Let |arrayBuffer| be |view|.\[[ViewedArrayBuffer]].
+1. Let |offset| be |view|.\[[ByteOffset]].
+1. Return |promise| and run the remaining steps [=in parallel=].
+1. [=stream/receive|Read bytes=] into |bytes|, from |internalStream|. Wait until either at least one
+   byte is read or FIN is received. Let |read| be the number of read bytes, and let |hasReceivedFIN|
+   be whether FIN was accompanied.
+
+
+   Note: The user-agent MAY have a buffer to improve the transfer performance. Such a buffer
+   SHOULD have a fixed size, to carry the backpressure information to the server.
+
+   Note: This operation may return before filling up all of |bytes|.
+
+1. If the previous step failed, abort the remaining steps.
+
+  Note: We don't reject |promise| here because we handle network errors elsewhere, and those logic
+  error |stream| and reject the result of this read operation.
+
+1. [=Queue a network task=] with |transport| to run these steps:
+
+   Note: If the buffer described above is available in the [=event loop=] where this procedure is
+   running, the following steps may run immediately.
+
+  1. Set |view| to a [=new=] {{Uint8Array}} with |arrayBuffer|, |offset| and |read|.
```

Comment:
https://github.com/whatwg/streams/pull/1146

Reply:
Hmm, the current logic uses _chunk_.[[ByteLength]] which is aligned with my understanding.

I'm worried about this assertion in the new logic:

>    1. Assert: if |bytesWritten| was given, it's equal to |chunk|.\[[ByteLength]]. (Anything else indicates a logic error in the calling specification.)

This requires WebTransport to predict the allocation size in front (which is possible for some but not all platforms), or recreate the view.

Reply:
> Hmm, the current logic uses chunk.[[ByteLength]] which is aligned with my understanding.

The idea is that your specification should not have to trim down the buffer, which incurs a copy. In other words the intent is to replace these two steps with

1. Enqueue |view| into |stream| with |read|.

> I'm worried about this assertion in the new logic:

I think recreating the view is necessary. The stream consumer should never receive a view which is padded with 0s. It's OK for the *buffer* to be padded with 0s, but not the view. The view's byte length should tell the consumer how many bytes were read.

Alternately, you could avoid allocating the view until |read| is known, and instead only allocate the underlying ArrayBuffer beforehand.

Reply:
> The idea is that your specification should not have to trim down the buffer, which incurs a copy.

Creating an ArrayBufferView from an ArrayBuffer doesn't lead to copying.

In the following code,

```
const ab = new ArrayBuffer(8);
let view1 = new Uint8Array(ab, 4, 3);
let view2 = new Uint8Array(ab, 4, 2);
```

_view1_ and _view2_ shares the same buffer (_ab_), and there is no copying when creating Uint8Arrays.
Am I missing something?

Reply:
You are right; I misspoke when I mentioned the copy. However it is redundant as ReadableByteStreamControllerRespond will eventually construct a new view itself. (In [ReadableByteStreamControllerConvertPullIntoDescriptor](https://streams.spec.whatwg.org/#readable-byte-stream-controller-convert-pull-into-descriptor).) You just need to pass the |read| so we know what size view to construct.

Reply:
I think this is now addressed - what do you think about the newest change?

Reply:
No, you are still manually constructing a view. My suggestion is the following:

Replace:

```
  1. If [=ReadableStream/current BYOB request view=] for |stream| is null:
    1. Let |arrayBuffer| be |view|.\[[ViewedArrayBuffer]].
    1. Let |offset| be |view|.\[[ByteOffset]].
    1. Set |view| to a [=new=] {{Uint8Array}} with |arrayBuffer|, |offset| and |read|.
    1. If |read| > 0, then [=ReadableStream/enqueue=] |view| into |stream|.
  1. Otherwise:
    1. If |read| > 0, then [=ReadableStream/enqueue=] |view| into |stream| with |read|.
```

with

```
  1. If |read| > 0, then [=ReadableStream/enqueue=] |view| into |stream| with |read|.
```

Then:

- If there was a BYOB request view back at the top of the algorithm, you are re-enqueuing that same view, but passing along |read|. The streams machinery will create a new view with the same backing buffer but with only |read| as its byte length.
- If there was no BYOB request view back at the top of the algorithm, you are enqueuing the new view you allocated at the top of the algorithm, but passing along |read|. The streams machinery will create a new view with the same backing buffer but with only |read| as its byte length, instead of "an implementation-defined size".

Reply:
That will be rejected with the assertion you added in whatwg/streams#1146. 

```
   1. Assert: if |bytesWritten| was given, it's equal to |chunk|.\[[ByteLength]]. (Anything else
      indicates a logic error in the calling specification.)
```

Reply:
You are right; sorry, I was very confused. I keep making mistakes on this :(

I think what I meant to suggest was the following: replace

```
1. Let |view| be [=ReadableStream/current BYOB request view=] for |stream|.
1. If |view| is null, set |view| to a [=new=] {{Uint8Array}} with an [=implementation-defined=]
   size. If allocating the {{Uint8Array}} fails, return a promise [=rejected with=] a
   {{RangeError}}.
1. Let |bytes| be the byte sequence |view| represents.
1. Assert: |bytes|'s length equals to |view|.\[[ByteLength]].
1. Return |promise| and run the remaining steps [=in parallel=].
1. [=ArrayBufferView/Write=] the bytes that area [=stream/receive|read=] from |internalStream| into
   |view|, up to |bytes|'s length bytes. Wait until either at least one byte is read or FIN is
   received. Let |read| be the number of read bytes, and let |hasReceivedFIN| be whether FIN was
   accompanied.
```

with

```
1. Let |buffer|, |offset|, and |maxBytes| be null.
1. If |stream|'s [=ReadableStream/current BYOB request view=] for |stream| is not null, then:
  1. Set |offset| to |stream|'s [=ReadableStream/current BYOB request view=].\[[ByteOffset]].
  1. Set |maxBytes| to |stream|'s [=ReadableStream/current BYOB request view=]'s [=BufferSource/byte length=].
  1. Set |buffer| to |stream|'s [=ReadableStream/current BYOB request view=]'s [=BufferSource/underlying buffer=].
1. Otherwise,
  1. Set |offset| to 0.
  1. Set |maxBytes| to an [=implementation-defined=] size.
  1. Set |buffer| be a [=new=] {{ArrayBuffer}} with |maxBytes| size. If allocating the {{Uint8Array}} fails, return a promise [=rejected with=] a {{RangeError}}.
1. Return |promise| and run the remaining steps [=in parallel=].
1. [=ArrayBuffer/Write=] the bytes that area [=stream/receive|read=] from |internalStream| into
   |buffer|, up to |maxBytes| bytes. Wait until either at least one byte is read or FIN is
   received. Let |read| be the number of read bytes, and let |hasReceivedFIN| be whether FIN was
   accompanied.
```

and then replace

```
  1. If [=ReadableStream/current BYOB request view=] for |stream| is null:
    1. Let |arrayBuffer| be |view|.\[[ViewedArrayBuffer]].
    1. Let |offset| be |view|.\[[ByteOffset]].
    1. Set |view| to a [=new=] {{Uint8Array}} with |arrayBuffer|, |offset| and |read|.
    1. If |read| > 0, then [=ReadableStream/enqueue=] |view| into |stream|.
  1. Otherwise:
    1. If |read| > 0, then [=ReadableStream/enqueue=] |view| into |stream| with |read|.
```

with

```
  1. If |read| > 0, then:
    1. Set |view| to a [=new=] {{Uint8Array}} with |buffer|, |offset| and |read|.
    1. [=ReadableStream/Enqueue=] |view| into |stream|.
```

WDYT?

Reply:
Thank you, sounds reasonable.

I accepted the suggestion with adding _offset_ at the reading step, with some minor style fixes.

With this we don't need whatwg/streams#1146, I think.